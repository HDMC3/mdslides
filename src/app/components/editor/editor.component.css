.nav-container {
    display: grid;
    grid-template-columns: 1fr 1fr;
    width: 100%;
}

.right-nav {
    display: flex;
    align-items: center;
    justify-content: right;
    gap: 10px;
}

#theme-button {
    display: none;
}

.sidebar-container {
    display: flex;
    flex-direction: column;
    align-items: center;
    gap: 1em;
}

#user-button {
    display: none;
}

.left-nav {
    display: flex;
    align-items: center;
    justify-content: left;
    gap: 10px
}

.title-presentation {
    width: 50%;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
}

.sidebar-container {
    display: flex;
    flex-direction: column;
    align-items: center;
    gap: 1em;
}

@media screen and (min-width: 500px) {
    #menu-button {
        display: none;
    }

    #theme-button {
        display: initial;
    }

    .title-presentation {
        width: 80%;
    }

    nb-sidebar {
        display: none;
    }
}