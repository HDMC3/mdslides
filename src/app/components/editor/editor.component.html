<nb-layout>

    <nb-layout-header fixed>
        <div class="nav-container">
            <div class="left-nav">
                <button id="menu-button" nbButton ghost (click)="toggleSidebar()">
                    <nb-icon icon="menu-outline"></nb-icon>
                </button>
                <h3 contenteditable>MDSlides</h3>
                <button nbButton shape="round">
                    <nb-icon icon="edit-outline"></nb-icon>
                </button>
            </div>
            <div class="right-nav">
                <button id="theme-button" nbButton shape="round" size="large" ghost (click)="changeTheme()">
                    <nb-icon [icon]="themeButtonIcon"></nb-icon>
                </button>
            </div>
        </div>
    </nb-layout-header>

    <nb-sidebar state="collapsed" fixed>
        <div class="sidebar-container">
            <button nbButton shape="round" size="large" ghost (click)="changeTheme()">
                <nb-icon [icon]="themeButtonIcon"></nb-icon>
            </button>
        </div>
    </nb-sidebar>

    <nb-layout-column>

    </nb-layout-column>

</nb-layout>