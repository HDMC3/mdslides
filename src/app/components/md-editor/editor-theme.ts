import { EditorView } from '@codemirror/view';

const tooltipBackground = '#353a42';

const textBasicColor = 'var(--text-basic-color)';
const textHintColor = 'var(--text-hint-color)';
const backgroundBasicColor1 = 'var(--background-basic-color-1)';
const backgroundBasicColor2 = 'var(--background-basic-color-2)';
const backgroundBasicColor4 = 'var(--background-basic-color-4)';
const colorBasic800 = 'var(--color-basic-800)';
const colorBasicTransparent600 = 'var(--color-basic-transparent-600)';
const colorControlDefault = 'var(--color-control-default)';
const colorControlActive = 'var(--color-control-active)';
const colorControlHover = 'var(--color-control-hover)';

export const oneDarkTheme = EditorView.theme({
    '&': {
        color: textBasicColor,
        backgroundColor: backgroundBasicColor1
    },

    '.cm-content': {
        caretColor: textHintColor
    },

    '.cm-cursor, .cm-dropCursor': { borderLeftColor: textHintColor },
    '&.cm-focused .cm-selectionBackground, .cm-selectionBackground, .cm-content ::selection': { backgroundColor: colorBasicTransparent600 },

    '.cm-panels': { backgroundColor: backgroundBasicColor4, color: textBasicColor },
    '.cm-panels.cm-panels-top': { borderBottom: '2px solid black' },
    '.cm-panels.cm-panels-bottom': { borderTop: '2px solid black' },

    '.cm-button': { border: 'none', backgroundImage: 'none', background: colorControlDefault, color: colorBasic800 },
    '.cm-button:active': { border: 'none', backgroundImage: 'none', background: colorControlActive },
    '.cm-button:hover': { border: 'none', backgroundImage: 'none', background: colorControlHover },
    '.cm-textfield': { color: colorBasic800 },

    '.cm-panel button[name=close]': { color: textBasicColor },

    '.cm-searchMatch': {
        backgroundColor: '#72a1ff59',
        outline: '1px solid #457dff'
    },
    '.cm-searchMatch.cm-searchMatch-selected': {
        backgroundColor: '#6199ff2f'
    },

    '.cm-activeLine': { backgroundColor: '#6699ff0b' },
    '.cm-selectionMatch': { backgroundColor: '#aafe661a' },

    '&.cm-focused .cm-matchingBracket, &.cm-focused .cm-nonmatchingBracket': {
        backgroundColor: '#bad0f847',
        outline: '1px solid #515a6b'
    },

    '.cm-gutters': {
        backgroundColor: backgroundBasicColor2,
        color: textHintColor,
        border: 'none'
    },

    '.cm-activeLineGutter': {
        backgroundColor: backgroundBasicColor4
    },

    '.cm-foldPlaceholder': {
        backgroundColor: 'transparent',
        border: 'none',
        color: '#ddd'
    },

    '.cm-tooltip': {
        border: 'none',
        backgroundColor: backgroundBasicColor4
    },
    '.cm-tooltip .cm-tooltip-arrow:before': {
        borderTopColor: 'transparent',
        borderBottomColor: 'transparent'
    },
    '.cm-tooltip .cm-tooltip-arrow:after': {
        borderTopColor: tooltipBackground,
        borderBottomColor: tooltipBackground
    },
    '.cm-tooltip-autocomplete': {
        '& > ul > li[aria-selected]': {
            backgroundColor: colorBasicTransparent600,
            color: textBasicColor
        }
    },
    '.ͼb': { color: 'var(--text-primary-focus-color)' },
    '.ͼc': {
        color: textHintColor
    },
    '.ͼd': { color: 'var(--text-warning-focus-color)' },
    '.ͼe': { color: 'var(--text-danger-focus-color)' },
    '.ͼf': { color: 'var(--text-danger-hover-color)' },
    '.ͼg': { color: '#940' },
    '.ͼh': { color: '#708' },
    '.ͼi': { color: 'var(--text-warning-color)' },
    '.ͼj': { color: 'var(--text-success-color)' },
    '.ͼk': { color: 'var(--text-success-focus-color)' },
    '.ͼl': { color: 'var(--text-info-color)' },
    '.ͼm': { color: 'var(--text-primary-color)' },
    '.ͼn': { color: 'var(--text-danger-color)' }
}, { dark: false });
