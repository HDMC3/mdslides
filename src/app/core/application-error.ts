export class ApplicationError extends Error {
    constructor(message: string, public title?: string) {
        super(message);
    }
}
