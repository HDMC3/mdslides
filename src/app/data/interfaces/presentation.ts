import { Slide } from './slide';

export interface Presentation {
    id: string;
    title: string;
    description: string;
    slides: Slide[];
    creation_date: Date;
    modification_date: Date;
}
