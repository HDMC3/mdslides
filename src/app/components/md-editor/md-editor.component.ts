import { Component, ElementRef, OnInit, Output, EventEmitter } from '@angular/core';
import { markdown } from '@codemirror/lang-markdown';
import { basicSetup } from 'codemirror';
import { EditorState } from '@codemirror/state';
import { languages } from '@codemirror/language-data';
import { EditorView } from '@codemirror/view';
import { oneDarkTheme } from './editor-theme';

@Component({
    selector: 'app-md-editor',
    templateUrl: './md-editor.component.html',
    styleUrls: ['./md-editor.component.css']
})
export class MdEditorComponent implements OnInit {
    @Output() valueChange: EventEmitter<string[]>;
    editorInitialized: boolean;

    constructor(private elementRef: ElementRef) {
        this.editorInitialized = false;
        this.valueChange = new EventEmitter();
    }

    ngOnInit(): void {
        const state = EditorState.create({
            extensions: [
                basicSetup,
                markdown({ codeLanguages: languages }),
                EditorView.domEventHandlers({
                    keyup: (e, v) => {
                        const editorValue = v.state.doc.toJSON();
                        this.valueChange.emit(editorValue);
                    }
                }),
                EditorView.lineWrapping,
                oneDarkTheme
            ]
        });

        const view = new EditorView({
            state,
            parent: this.elementRef.nativeElement
        });

        view.dom.style.height = '100%';
    }

    initEditor() {
    }

}
