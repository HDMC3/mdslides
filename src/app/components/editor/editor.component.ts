import { Component, ElementRef, OnInit, ViewChild, AfterViewInit, HostListener, OnDestroy } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { NbDialogService, NbSidebarService, NbThemeService } from '@nebular/theme';
import { take } from 'rxjs/operators';
import { Presentation } from 'src/app/data/interfaces/presentation';
import { EditTitleDialogComponent } from '../edit-title-dialog/edit-title-dialog.component';

@Component({
    selector: 'app-editor',
    templateUrl: './editor.component.html',
    styleUrls: ['./editor.component.css']
})
export class EditorComponent implements OnInit, AfterViewInit, OnDestroy {
    @ViewChild('divider') divider?: ElementRef;
    @ViewChild('mdEditor') mdEditor?: ElementRef;
    @ViewChild('currentSlide') currentSlide?: ElementRef;
    currentTheme: string;
    themeButtonIcon: string;
    presentation: Presentation;
    isClickDivider: boolean;
    isMdEditorActive: boolean;
    resizeObserver: ResizeObserver;

    constructor(
        private nbThemeService: NbThemeService,
        private nbSidebarService: NbSidebarService,
        private activatedRoute: ActivatedRoute,
        private nbDialogService: NbDialogService
    ) {
        this.currentTheme = localStorage.getItem('nbTheme') ?? 'default';
        this.themeButtonIcon = this.currentTheme === 'default' ? 'moon-outline' : 'sun-outline';
        this.nbThemeService.changeTheme(this.currentTheme);

        this.presentation = {
            id: '',
            title: 'Mi titulo',
            description: '',
            slides: [],
            creation_date: new Date(),
            modification_date: new Date()
        };

        this.isClickDivider = false;
        this.isMdEditorActive = false;

        this.resizeObserver = new ResizeObserver(() => {
            if (window.innerWidth <= 500 && this.mdEditor && this.currentSlide) {
                this.mdEditor.nativeElement.style = '';
                this.currentSlide.nativeElement.style = '';
            }
        });
    }

    ngOnInit(): void {
        this.activatedRoute.params.pipe(take(1)).subscribe(params => {
            this.presentation.id = params['id'];
        });
    }

    ngAfterViewInit(): void {
        this.resizeObserver.observe(document.body);
    }

    ngOnDestroy(): void {
        this.resizeObserver.disconnect();
    }

    @HostListener('pointermove', ['$event'])
    onMouseMoveDivider(e: PointerEvent) {
        if (!this.isClickDivider) return;

        if (!this.mdEditor || !this.currentSlide || !this.divider) return;

        if (window.innerWidth > 500) {
            const offset = e.clientX - this.divider.nativeElement.offsetLeft;
            const newCurrentSlideWidth = this.currentSlide.nativeElement.clientWidth - offset;
            const newMdEditorWidth = this.mdEditor.nativeElement.clientWidth + offset;
            this.currentSlide.nativeElement.style.width = `${newCurrentSlideWidth}px`;
            this.mdEditor.nativeElement.style.width = `${newMdEditorWidth}px`;
        }
    }

    @HostListener('pointerdown', ['$event'])
    onPointerDownDivider(e: any) {
        this.isClickDivider = e.target && e.target.classList.contains('editor-section-divider');
    }

    @HostListener('pointerup', ['$event'])
    onPointerUpDivider(e: any) {
        this.isClickDivider = false;
    }

    switchView() {
        if (!this.mdEditor || !this.currentSlide) return;

        if (!this.isMdEditorActive) {
            this.currentSlide.nativeElement.classList.add('hidde-panel');
            this.mdEditor.nativeElement.classList.remove('hidde-panel');
            this.isMdEditorActive = true;
            return;
        }

        this.mdEditor.nativeElement.classList.add('hidde-panel');
        this.currentSlide.nativeElement.classList.remove('hidde-panel');
        this.isMdEditorActive = false;
    }

    changeTheme() {
        const newTheme = this.currentTheme === 'default' ? 'dark' : 'default';
        localStorage.setItem('nbTheme', newTheme);
        this.currentTheme = newTheme;
        this.nbThemeService.changeTheme(newTheme);
        this.themeButtonIcon = newTheme === 'default' ? 'moon-outline' : 'sun-outline';
    }

    toggleSidebar() {
        this.nbSidebarService.toggle();
    }

    openEditTitleDialog() {
        this.nbDialogService.open(EditTitleDialogComponent, {
            closeOnBackdropClick: true,
            context: {
                presentation: this.presentation
            }
        });
    }

}
