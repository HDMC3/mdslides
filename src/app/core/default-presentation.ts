import { v4 } from 'uuid';
import { Presentation } from '../data/interfaces/presentation';

export const getDefaultPresentation = (id: string): Presentation => {
    return {
        id,
        title: 'Titulo de presentacion',
        creation_date: new Date(),
        modification_date: new Date(),
        description: '',
        slides: [
            {
                id: v4(),
                name: 'Presentacion inicial',
                code: [
                    '---',
                    'style: |',
                    'section {',
                    '   display: flex;',
                    '   flex-direction: column;',
                    '   align-items: center;',
                    '   justify-content: center;',
                    '}',
                    '---',
                    '',
                    '# Inicio de la presentacion',
                    '# ðŸ˜„âœŒ'
                ]
            }
        ]
    };
};
