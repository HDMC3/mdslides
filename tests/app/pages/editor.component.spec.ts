import { ComponentFixture, TestBed } from '@angular/core/testing';
import { NbDialogModule, NbSidebarModule, NbThemeModule, NbToastrModule } from '@nebular/theme';
import { MdEditorService } from '../../../src/app/core/services/md-editor.service';
import { PresentationFileService } from '../../../src/app/core/services/presentation-file.service';
import { PresentationService } from '../../../src/app/core/services/presentation.service';
import { ThemeService } from '../../../src/app/shared/services/theme.service';

import { EditorComponent } from '../../../src/app/pages/editor/editor.component';
import { RouterTestingModule } from '@angular/router/testing';

window.ResizeObserver =
    window.ResizeObserver ||
    jest.fn().mockImplementation(() => ({
        disconnect: jest.fn(),
        observe: jest.fn(),
        unobserve: jest.fn()
    }));

describe('EditorComponent', () => {
    let component: EditorComponent;
    let fixture: ComponentFixture<EditorComponent>;
    let compiled: HTMLElement;

    beforeEach(async() => {
        await TestBed.configureTestingModule({
            imports: [
                NbThemeModule.forRoot({ name: 'default' }),
                NbSidebarModule.forRoot(),
                NbDialogModule.forRoot(),
                NbToastrModule.forRoot(),
                RouterTestingModule.withRoutes([])
            ],
            providers: [
                MdEditorService,
                ThemeService,
                PresentationService,
                PresentationFileService
            ]
        })
            .compileComponents();
    });

    beforeEach(() => {
        fixture = TestBed.createComponent(EditorComponent);
        component = fixture.componentInstance;
        fixture.detectChanges();
        compiled = fixture.nativeElement;
    });

    it('should create', () => {
        expect(component).toBeTruthy();
    });

    it('should match the snapshot', () => {
        expect(compiled).toMatchSnapshot();
    });
});
