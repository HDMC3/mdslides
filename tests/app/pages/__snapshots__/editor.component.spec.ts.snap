// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditorComponent should match the snapshot 1`] = `
<div
  id="root1"
>
  <nb-layout>
    <nb-layout-header
      fixed=""
    >
      <div
        class="nav-container"
      >
        <div
          class="left-nav"
        >
          <button
            ghost=""
            id="menu-button"
            nbbutton=""
          >
            <nb-icon
              icon="menu-outline"
            />
          </button>
          <button
            ghost=""
            nbbutton=""
            shape="round"
          >
            <nb-icon
              icon="edit-outline"
            />
          </button>
          <h4
            class="title-presentation"
          >
            Titulo de presentacion
          </h4>
        </div>
        <div
          class="right-nav"
        >
          <button
            ghost=""
            id="theme-button"
            nbbutton=""
            shape="round"
            size="large"
          >
            <nb-icon />
          </button>
          <button
            id="new-presentation-button"
            nbbutton=""
          >
            <nb-icon
              icon="file"
            />
          </button>
          <button
            id="open-presentation-button"
            nbbutton=""
            status="warning"
          >
            <nb-icon
              icon="folder"
            />
          </button>
          <button
            id="save-presentation"
            nbbutton=""
            status="info"
          >
            <nb-icon />
          </button>
          <button
            id="show-presentation-button"
            nbbutton=""
            status="success"
          >
            Presentar 
            <nb-icon
              icon="expand"
            />
          </button>
        </div>
      </div>
    </nb-layout-header>
    <nb-sidebar
      fixed=""
      state="collapsed"
    >
      <div
        class="sidebar-container"
      >
        <button
          fullwidth=""
          nbbutton=""
          status="success"
        >
          Presentar 
          <nb-icon
            icon="expand"
          />
        </button>
        <button
          fullwidth=""
          nbbutton=""
          status="info"
        >
           Guardar 
          <nb-icon />
        </button>
        <button
          fullwidth=""
          nbbutton=""
          status="warning"
        >
           Abrir 
          <nb-icon
            icon="folder"
          />
        </button>
        <button
          fullwidth=""
          nbbutton=""
        >
           Nueva 
          <nb-icon
            icon="file"
          />
        </button>
        <button
          ghost=""
          nbbutton=""
          shape="round"
          size="large"
        >
          <nb-icon />
        </button>
      </div>
    </nb-sidebar>
    <nb-layout-column>
      <main>
        <section
          class="editor-section"
        >
          <div
            class="md-editor hidde-panel"
          >
            <app-md-editor />
          </div>
          <div
            class="editor-section-divider"
          />
          <app-current-slide />
          <button
            id="switch-view-button"
            nbbutton=""
            status="info"
          >
            <nb-icon
              icon="swap"
            />
          </button>
        </section>
        <app-miniatures />
      </main>
    </nb-layout-column>
  </nb-layout>
</div>
`;
